ppa/archive/gale.py

```diff
@@ -0,0 +1,173 @@
+
+        # Returns the response for status 200 OK; raises
+        # :class:`HathiItemNotFound` for 404 and :class:`HathiItemForbidden`
+        # for 403.
+        # '''
+        rqst_url = "%s/%s" % (self.api_root, url)
+        rqst_opts = {}
+        if params:
+            rqst_opts["params"] = params.copy()
+
+        # add api key to parameters if neded for this request
+        if requires_api_key:
+            if "params" not in rqst_opts:
+                rqst_opts["params"] = {}
+            rqst_opts["params"]["api_key"] = self.api_key
+
+        resp = self.session.get(rqst_url, stream=stream, **rqst_opts)
+        logger.debug(
+            "get %s %s: %f sec", rqst_url, resp.status_code, resp.elapsed.total_seconds()
+        )
+        if resp.status_code == requests.codes.ok:
+            return resp
+        if resp.status_code == requests.codes.not_found:
+            raise GaleItemNotFound
+
+        # when api key expires, API returns:
+        # HTTP Status 401 - Authentication Failed: Invalid or Expired API key
+        # If we get a 401 on a request that requires an api key, try getting a new one
+        if resp.status_code == requests.codes.unauthorized:
+            # If we get a 401 on a request that requires an api key,
+            # get a fresh key and then try the same request again
+            if requires_api_key and retry < 1:
+                self.refresh_api_key()
+                return self._make_request(
+                    url,
+                    params=params,
+                    requires_api_key=requires_api_key,
+                    stream=stream,
+                    retry=retry + 1
+                )
+            # response is html error, not json; could try
+            # extracting h1, but not sure it's worth parsing
+            raise GaleUnauthorized()
+
+        if resp.status_code == requests.codes.forbidden:
+            # forbidden results return a message
+            # NOTE that item requests for invalid ids may return 403
+            raise GaleItemForbidden(resp.json()["message"])
+
+    def get_api_key(self):
+        """Get a new API key to use for requests in the next 30 minutes."""
+        # GALE API requires use of an API key, which lasts for 30 minutes
+        # request a new one when needed using configured username
+        response = self._make_request(
+            "tools/generate_key", {"user": self.username}, requires_api_key=False
+        )
+        return response.json()["apiKey"]
+
+    _api_key = None
+
+    @property
+    def api_key(self):
+        """Property for current api key. Uses :meth:`get_api_key` to
+        request a new one when needed."""
+        if self._api_key is None:
+            self._api_key = self.get_api_key()
+        return self._api_key
+
+    def refresh_api_key(self):
+        """clear cached api key and request a new one"""
+        self._api_key = None
+        assert self.api_key   # populate new key through property
```

Comment:
whoa, the usage of `assert` here is really interesting. took me a second to understand, but I guess it might be the most concise way to accomplish this!

Reply:
I had it without the `assert` at first and one of the code scanners complained that it wasn't doing anything (even though I think it would work without); thought this was the simplest solution. (technically might not even need this since `_make_request` uses it by the property, which would populate if not set...)